pkg/store/bucket.go

```diff
@@ -2325,112 +2321,228 @@ func ExpandPostingsWithContext(ctx context.Context, p index.Postings) (res []sto
 	return res, p.Err()
 }
 
-// postingGroup keeps posting keys for single matcher. Logical result of the group is:
+// postingGroup keeps posting keys for one or more matchers with the same label name. Logical result of the group is:
 // If addAll is set: special All postings minus postings for removeKeys labels. No need to merge postings for addKeys in this case.
 // If addAll is not set: Merge of postings for "addKeys" labels minus postings for removeKeys labels
 // This computation happens in ExpandedPostings.
 type postingGroup struct {
 	addAll     bool
-	addKeys    []labels.Label
-	removeKeys []labels.Label
+	name       string
+	addKeys    []string
+	removeKeys []string
 }
 
-func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
+func newPostingGroup(addAll bool, name string, addKeys, removeKeys []string) *postingGroup {
 	return &postingGroup{
 		addAll:     addAll,
+		name:       name,
 		addKeys:    addKeys,
 		removeKeys: removeKeys,
 	}
 }
 
-func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
+func (pg postingGroup) merge(other *postingGroup) *postingGroup {
+	if other == nil {
+		return &pg
+	}
+	// This shouldn't happen, but add this as a safeguard.
+	if pg.name != other.name {
+		return nil
+	}
+	// Both add all, merge remove keys.
+	if pg.addAll && other.addAll {
+		// If nothing to merge, end early.
+		if len(pg.removeKeys) == 0 && len(other.removeKeys) == 0 {
+			return &pg
+		}
+		if len(other.removeKeys) == 0 {
+			return &pg
+		}
+		if len(pg.removeKeys) == 0 {
+			pg.removeKeys = other.removeKeys
+			return &pg
+		}
+		set := make(map[string]struct{}, len(pg.removeKeys)+len(other.removeKeys))
+		for _, key := range pg.removeKeys {
+			set[key] = struct{}{}
+		}
+		for _, key := range other.removeKeys {
+			set[key] = struct{}{}
+		}
+		pg.removeKeys = make([]string, len(set))
+		i := 0
+		for key := range set {
+			pg.removeKeys[i] = key
+			i++
+		}
+	} else if pg.addAll || other.addAll {
+		// Subtract the remove keys.
+		toRemove := other
+		toAdd := &pg
+		if pg.addAll {
+			toRemove = &pg
+			toAdd = other
+		}
+		set := make(map[string]struct{}, len(toRemove.removeKeys))
+		for _, key := range toRemove.removeKeys {
+			set[key] = struct{}{}
+		}
+		i := 0
+		for _, key := range toAdd.addKeys {
+			if _, ok := set[key]; !ok {
+				toAdd.addKeys[i] = key
+				i++
+			}
+		}
+		toAdd.addKeys = toAdd.addKeys[:i]
+		if pg.addAll {
+			pg.removeKeys = nil
+			pg.addKeys = other.addKeys
+			pg.addAll = false
+		}
+	} else {
+		set := make(map[string]struct{}, len(pg.addKeys))
+		for _, key := range pg.addKeys {
+			set[key] = struct{}{}
+		}
+		mergedKeys := make([]string, 0)
+		for _, key := range other.addKeys {
+			if _, ok := set[key]; ok {
+				mergedKeys = append(mergedKeys, key)
+			}
+		}
+		pg.addKeys = mergedKeys
```

Comment:
Should we merge the removeKeys as well just in case? I know this should not happen if its not AddAll from the code but we could do it just for the sake of correctness?

Reply:
Discussed and seems this is okay to skip